Proxy pass through for Babbage

MariliaE opened the pull request:
### What
A proxy handler had been created to handle all HTTP requests made from dp-frontend-router to Babbage and all HTTP responses from Babbage to the router.

### How to review

1. Run tests `make test`
2. In the dp-frontend-router checkout the branch of [this PR](https://github.com/ONSdigital/dp-frontend-router/pull/444)
Set _LEGACY_CACHE_PROXY_ENABLED_ to true by running this on the terminal
`export LEGACY_CACHE_PROXY_ENABLED=true` and run dp-frontend-router using `make debug`
4. Run dp-legacy-cache-proxy `make debug`
5. Run babbage `./run.sh`
6. Run zebedee `./run-reader.sh` (requires content to have been generated as described in the zebedee README.md prior to this)
7. Make a request to a Babbage endpoint
e.g. http://localhost:20000/economy/environmentalaccounts/bulletins/ukenvironmentalaccounts/2013-06-26

8. HTML content should be returned 

### Who can review

Anyone from ONS

jmgq commented on proxy/handler.go:
```diff
@@ -0,0 +1,52 @@
+	"github.com/ONSdigital/log.go/v2/log"
+)
+
+func (proxy *Proxy) Handler(ctx context.Context, w http.ResponseWriter, req *http.Request, cfg *config.Config) {
```
It looks like the only thing that the config is being used for is to get the Babbage URL, so you can simply pass that instead.

jmgq replied:
Another thing is that this file name and this method name should probably be changed to something else, since this is not really a handler, and it does not return a handler either. I'm not sure what the right name could be... So far, this is a method that passes the request to Babbage, but later on this will be the central place while the whole proxy logic occurs.

jmgq commented on proxy/handler.go:
```diff
@@ -0,0 +1,52 @@
+func (proxy *Proxy) Handler(ctx context.Context, w http.ResponseWriter, req *http.Request, cfg *config.Config) {
+	targetURL := cfg.BabbageURL + req.URL.String()
+
+	client := &http.Client{}
```
I'd move this declaration closer to where it's being used (above `client.Do(proxyReq)`). For two reasons:
- Them being together makes the code easier to read.
- The next block does not use the client, so if `err` is not `nil`, then we would have created this variable unnecessarily.

jmgq commented on proxy/handler.go:
```diff
@@ -0,0 +1,52 @@
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
```
There is an unhandled error here (`Close` can return an error).

MariliaE replied:
Would for this one make sense to just log the closing error without returning an http error back to the dp-frontend-router? By the time it reaches this part of the code it means we've already read the response and wrote it back to the client without any issues.

jmgq replied:
If that's the case, then yeah, that makes sense, thanks!

jmgq commented on proxy/handler.go:
```diff
@@ -0,0 +1,52 @@
+
+	proxyReq, err := http.NewRequestWithContext(ctx, req.Method, targetURL, req.Body)
+	if err != nil {
+		log.Error(ctx, "proxyHandler: Error constructing proxy request", err)
```
Just a reminder that all these error logs would need refactoring when you change this method's name.

jmgq commented on proxy/handler.go:
```diff
@@ -0,0 +1,52 @@
+	proxyReq, err := http.NewRequestWithContext(ctx, req.Method, targetURL, req.Body)
+	if err != nil {
+		log.Error(ctx, "proxyHandler: Error constructing proxy request", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
By passing `err.Error()`, doesn't that mean that the end user will be able to see that message? If that's the case, we need to change this everywhere, because we don't want to leak implementation details to the user. A generic "Internal Server Error" would be enough. The specific error is useful for ONS, and you're already logging it in the line above.

jmgq commented on proxy/proxy_test.go:
```diff
@@ -14,12 +16,14 @@ func TestSetup(t *testing.T) {
-		Convey("When created the following routes should have been added", func() {
-			// Replace the check below with any newly added routes
-			So(hasRoute(proxy.Router, "/hello", "GET"), ShouldBeTrue)
+		Convey("When created the following routes should be available", func() {
```
You could reword this so that it reflects that all methods are allowed (and then add the missing methods below).

jmgq commented on proxy/handler_test.go:
```diff
@@ -0,0 +1,60 @@
+		router := mux.NewRouter()
+		ctx := context.Background()
+
+		Convey("when sending a proxy request to Babbage", func() {
```
```suggestion
		Convey("When sending a proxy request to Babbage", func() {
```

jmgq commented on proxy/handler_test.go:
```diff
@@ -0,0 +1,60 @@
+			})
+		})
+
+		Convey("when there is an error with sending the proxy request to Babbage", func() {
```
```suggestion
		Convey("When there is an error with sending the proxy request to Babbage", func() {
```

jmgq replied:
Also, this message is misleading. What the message says would actually be a good test, but that's not what you're actually testing. You're testing that when Babbage returns a 500 error, the proxy also returns that. But this is simply an extension of the test that you've already written above. What you should be testing here is the different scenarios in which the proxy itself throws an error (for instance, when `http.NewRequestWithContext` or `client.Do(proxyReq)` or `io.Copy` fail).

MariliaE replied:
Testing `http.NewRequestWithContext` and `io.Copy` turned out to be difficult as we couldn't get them to error through the tests. The only other option we thought of is to mock the http client for `http.NewRequestWithContext ` and for `io.Copy` to simulate a custom Writer that returns an error and would cause it to fail. However both of these seemed a bit unclear and convoluted, would you have any suggestions?

jmgq replied:
I think if some of these errors are hard to reproduce, we're OK not covering them, as they're not really part of the business logic anyway. They're there just as a safeguard.

jmgq commented on proxy/handler_test.go:
```diff
@@ -0,0 +1,60 @@
+var babbageEndpoint = "/economy/environmentalaccounts/bulletins/ukenvironmentalaccounts/2013-06-26"
+
+func TestProxyHandler(t *testing.T) {
+	Convey("Given a Proxy Handler ", t, func() {
```
This is called "Given a Proxy Handler", but that's not really what's going on here, you've only defined a router and a context. In any case, when writing tests in the Given/When/Then format, you should try to organise them so that the setup happens in the Given block (for instance, you could move the creation of the `legacyProxy` variable here), the execution happens in the When block, and the assertions happens in the Then block. This isn't a hard rule, sometimes it may be necessary to mix things a bit, or it may be unpractical to be that rigid, but in general you should adhere to this, as it's a well known convention.

MariliaE replied:
Yes, this makes sense and that's how I had it originally, but due to needing a different configuration in each proxy setup for mocking the babbage server I moved this bit of the code within the Given blocks

jmgq commented on proxy/handler_test.go:
```diff
@@ -0,0 +1,60 @@
+
+			legacyProxy.Router.ServeHTTP(w, r)
+
+			Convey("Then return the Babbage response with status code 200", func() {
```
This isn't right. The only reason we're returning a status code of 200 is because that's what Babbage returns, but if it had returned a code of 404 (for instance), we should also return that. There are other things that our proxy does that are not being tested anywhere. For instance, you're not checking for the headers anywhere.

jmgq commented on proxy/handler_test.go:
```diff
@@ -0,0 +1,60 @@
+		ctx := context.Background()
+
+		Convey("when sending a proxy request to Babbage", func() {
+			babbageMockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
I think you're supposed to call `Close` after you're finished with the server.

jmgq committed:
Implement a No-Operation proxy

The proxy now accepts any incoming requests, forward them to Babbage,
and returns the Babbage response unmodified.
